website/docs/releases/migration-4.md

```diff
@@ -21,6 +21,16 @@ module.exports = {
 }
 ```
 
+### I18nProvider no longer remounts its children on locale change
+
+Previously, the `I18nProvider` remounted its children on locale change. This ensured that the whole app was re-rendered with the new locale and all strings were rendered correctly translated.
+Apart from not being very performant, this approach had the drawback that the state of the app was lost - all components were re-mounted and their state was reset. This is not a standard behavior of React Context providers and could cause some confusion.
+
+In v4, the `I18nProvider` no longer remounts its children on locale change. Instead, when locale changes, the context value provided by `I18nProvider` is updated and all components that consume the provided React Context are re-rendered with the new locale.
+This includes components provided by Lingui, such as `Trans` or `Plural` and also custom components that use the `useLingui` hook.
+
+This is the default behavior, but you can disable it by setting `forceRenderOnLocaleChange={false}`.
```

Comment:
is there any way to turn v3 behaviour for transition period? As far as i understand, `forceRenderOnLocaleChange={false}` will not make it as in v3, right?

Reply:
I checked the code one more time and question arised. What is the value of `forceRenderOnLocaleChange=false` ? I could not imagine a usecase where user will not want to have application to react to the locale changes. 

I think the previous behaviour of this flag was indeed confusing and some people wanted to disable it, that's why this flag exists. But with current implementation, is it better to delete it completely?  Or make this flag kinda enabling v3 compatibility?

Reply:
hey @thekip I'm not sure about the history of the code and why `forceRenderOnLocaleChange` was added TBH. 
I guess _someone_ had _some_ use case for the `I18nProvider` to not do anything "clever" but only pass the i18n object down.

I think a good way forward would be to remove the `forceRenderOnLocaleChange` prop, have it behave as if `forceRenderOnLocaleChange={true}` AND export the `LinguiContext` so that people can build their own context providers where they can do whatever they want.
This will make usual things easy to do, and harder / unusual things possible (the v3 behavior, for example - whatever it was 😄).

Wdyt?

Reply:
https://github.com/lingui/js-lingui/issues/1194 - might be related

Reply:
@vonovak 
> export the LinguiContext so that people can build their own context providers

I always thought that exporting context directly from the lib is not a good thing. At least i haven't seen such examples in big projects. 

Usually, libs give more high level primitives to work with. Such as `usLingui()` hook. What could be done with exported Context what couldn't be done with just `useLingui` hook? (sorry, I'm not too much into React)

> I think a good way forward would be to remove the forceRenderOnLocaleChange prop

What stopping us from deleting it in this PR?  I don't see a value of this anymore.  

I also a bit worried about transition path. If someone rely on the `forceRenderOnLocaleChange` in v3, for transition to v4, should go and find all places where unsafe `t` (and other from `core`) macro calls are used and add a useLingui hook. That might be quite a blocking factor for adopting v4.  What do you think?

Reply:
> I always thought that exporting context directly from the lib is not a good thing. At least i haven't seen such examples in big projects.

Would be nice to not have to do that but I don't think it does a lot of harm (it's just an exported object) and it gives users the flexibility to do _anything_ they want.

Here are some examples that export the Context: [react-redux](https://github.com/reduxjs/react-redux/blob/8d03182d36abe91cb0cc883478f3b0c2d7f9e17f/src/exports.ts#L54), [mobx-react](https://github.com/mobxjs/mobx/blob/01432751d235da15fc861619563a08f1c9dfa623/packages/mobx-react/src/index.ts#L21) - these are pretty big projects.

> Usually, libs give more high level primitives to work with

yes, and we give the `I18nProvider` which should cover the usual cases. And `useLingui()` which works fine as-is.

If our `I18nProvider` does not suit someone well, and the `LinguiContext` is exported, we're not blocking users to do whatever they please. When someone encounters a problem with `I18nProvider` and opens an issue, a solution _can_ be found if the `LinguiContext` can be used by developers (Then they can share their custom `I18nProvider` and perhaps we can integrate it in the next version).

If we don't export the `LinguiContext` then we're blocking people from making custom solutions possible. At the same time, exporting is doesn't mean maintenance problems for the lib afaict.

> What could be done with exported Context what couldn't be done with just useLingui hook?

Well, this whole question concerns the `I18nProvider`, not really the `useLingui` hook which works fine. But to answer "what can be done with with exported Context". Well, the answer is that anything can be done. People use opens source software in ways that a maintainer cannot always anticipate and I think it's best not to block people from doing what they want unless it has maintenance burden (exporting one Context object is not a maintenance burden).

If you want a more specific example I think I can come up with something :).

> What stopping us from deleting `forceRenderOnLocaleChange` in this PR?

Nothing, I'll happily do it.


> If someone rely on the forceRenderOnLocaleChange in v3, for transition to v4, should go and find all places where unsafe t (and other from core) macro calls are used and add a useLingui hook. That might be quite a blocking factor for adopting v4.

they should do it anyways, regardless of whether they had `forceRenderOnLocaleChange` true or false.

If they (incorrectly) used the `t` macro without `useLingui` then their app _might_ break in some cases. If we export `LinguiContext` then they can copy the `I18nProvider` from v3 into their code base. Then they will have the old behavior of translation rendering but enjoy other benefits of v4.
I agree I need to document this better.

Reply:
@vonovak can we move forward with this PR and merge it or you're going to address the discussion regarding the `forceRenderOnLocaleChange` here?